### PR TITLE
chore(#9080): adds logger shared lib

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

#9080

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8338-environment-shared-lib/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8338-environment-shared-lib/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8338-environment-shared-lib/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: shared-libs/logger/src/index.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
+if (!process || process.browser) {
+  module.exports = console;
+} else {
```
Comment (garethbowen) [inline shared-libs/logger/src/index.js:3]: Having different `module.exports` code depending on this branch makes it really hard to work out what it's going to do, especially if you're calling it from another shared lib and you don't know what context you're in. I think it'd be easier to read if it was something like...

```js
const create() => {
  if (!process || process.browser) {
    return console;
  } else {
    return require('./node-logger.js').create();
  }
};

module.exports = create();
module.exports.DATE_FORMAT = DATE_FORMAT;
```

File: shared-libs/outbound/src/outbound.js [hunk 2]
```diff
@@ -280,41 +278,37 @@ const alreadySent = (payload, configName, recordInfo) => {
...
-            .then(() => logger.info(`Pushed ${record._id} to ${configName}`))
-            .then(() => true);
-        })
-        .catch(err => {
-          logSendError(configName, record._id, err);
-          throw err;
-        });
-    },
-  };
+module.exports = {
```
Comment (garethbowen) [inline shared-libs/outbound/src/outbound.js:281]: :heart:

File: shared-libs/transitions/test/unit/transitions/mark_for_outbound.js [hunk 3]
```diff
@@ -1,19 +1,19 @@
-const rewire = require('rewire');
```
Comment (garethbowen) [inline shared-libs/transitions/test/unit/transitions/mark_for_outbound.js:1]: Woohoo! Death to rewire!

**Review by garethbowen — CHANGES_REQUESTED** [review]
One refactor suggestion, but otherwise this is great!

File: shared-libs/transitions/test/unit/transitions/mark_for_outbound.js (hunk 3 above)
Comment (dianabarsan) [reply shared-libs/transitions/test/unit/transitions/mark_for_outbound.js:1]: yes! i concur!

**Comment by garethbowen** [comment]
@dianabarsan One more random thought... in node services we often log like `log.error("got error: %o", err)` - but that won't work in the browser, right? Do we have to just be careful to not use that format in shared-libs?

**Comment by dianabarsan** [comment]
> in node services we often log like log.error("got error: %o", err)

Correct, this kind of formatting won't work in the browser. Nothing is going to break, but we won't achieve the desired result. I suppose we can continue to use this format when we know the code will only be used in nodejs - in libraries like server-checks for example.

**Comment by garethbowen** [comment]
@dianabarsan What about this: https://www.npmjs.com/package/winston-transport-browserconsole ? It might mean the logging API is consistent between node and browser?

**Comment by dianabarsan** [comment]
> @dianabarsan What about this: https://www.npmjs.com/package/winston-transport-browserconsole ? It might mean the logging API is consistent between node and browser?

I've read about it. People said it bulks the browser bundle. I'll try it out though.

**Comment by dianabarsan** [comment]
It looks like `winston-transport-browserconsole` it doesn't work out of the box:

image

It looks like `setImmediate` has been deprecated for a while: https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate and the library hasn't received any update in years. I'm reluctant to use it. 

Do you think it's worth more time working out how to generalize logging in webapp in this PR?

**Review by garethbowen — APPROVED** [review]
Awesome, thanks!

**Pull request merged**